### ISSUE TITLE
Encrypt all docs that are stored in OpenEMR on drive (ie. mostly patient docs and some logs)

### ISSUE THREAD

**Issue by bradymiller** [openemr/openemr#2181]
This will be a fun project. Will include a couple steps:
1. build out current encryption to create/store keys in a new sql table (we currently have a mechanism that stores keys on the drive to encrypt stuff in database, so just need to leverage that a little bit).
2. encrypt/decrypt the billing log as proof of concept
3. then go wild and encrypt/decrypt mechanism for patient documents
4. then do the miscellaneous stuff
(note we can turn off/on this feature via a global, and store whether docs/logs are encrypted/decrypted in the sql entries)
Some may ask, why do this. Just seem to make sense since we now rely on apache to not let anyone get to the documents directory; having this would give another line of defense in case a user does not set up apache properly or if a hacker gets through it.

**Comment by bradymiller** [comment]
Note I am jump starting this project a bit (did steps 1-2) in the following PR:
https://github.com/openemr/openemr/pull/2211

**Comment by bradymiller** [comment]
Steps 1 and 2 are completed. Here are 2 additional projects that can be now completed:

1. Encrypt/decrypt the letters module files that are stored on drive (small project)

2. Encrypt/decrypt the documents module that are stored on drive (large project)

If anybody has interest in these projects, just let me know and then I'll create an issue with more information/guidance on getting started.

**Comment by prondubuisi** [comment]
@bradymiller  I would love to work on this, you can create the issue with more information/guidance on getting started

**Comment by danehrlich1** [comment]
@bradymiller Can you provide @prondubuisi with links to the "patient letters module" and the "patient documents module". I don't even know where these are in the codebase sadly.

Also please provide a link to the proof of concept file you did.

**Comment by bradymiller** [comment]
hi @prondubuisi and @danehrlich1 ,

For proof of concept, see the current example where doing encryption of the billing log that is stored on drive.
The encryption is here: https://github.com/openemr/openemr/blob/master/interface/billing/billing_process.php#L310-L315
And the decryptions are here:
https://github.com/openemr/openemr/blob/master/interface/billing/billing_process.php#L158-L161
https://github.com/openemr/openemr/blob/master/interface/billing/customize_log.php#L19-L23

Lets work on the letter module first.
The patient letter module can be seen in OpenEMR. Open a patient, and then in top menu go to popups->Letter. This script is at https://github.com/openemr/openemr/blob/master/interface/patient_file/letter.php . Recommend reading through the script to see how it writes to the drive. Goal it to encrypt it when write it to the drive and decrypt when read it from the drive.

-brady

**Comment by prondubuisi** [comment]
Hello @bradymiller I am working on this but I can't see any new letters in the folder, am guessing its ubuntu permission errors, How can i turn on error reporting, I installed Openemr using docker and I see plenty php.ini files for different PHP version folders in the dockers folder

**Comment by bradymiller** [comment]
hi @prondubuisi , very nice to see you working on this. pretty excited about this issue. Within the docker, the log will be at /var/log/apache2/error.log . Let us know if you need guidance on how to get into the docker. -brady

**Comment by prondubuisi** [comment]
@bradmiller I can't see any /var folder, at least not in the root directory. How do I access it?

**Comment by bradymiller** [comment]
what do you see when do:
```
docker ps -a
```

**Comment by bradymiller** [comment]
@prondubuisi , forgot to flag you on my post above

**Comment by prondubuisi** [comment]
okay @bradymiller this is what I see from the command https://pastebin.com/M191WDvc

**Comment by bradymiller** [comment]
hi @prondubuisi ,
To go into the docker, try:
```
docker exec -it openemr_openemr_1 sh
```
then in the docker, try:
```
cat /var/log/apache2/error.log
```

**Comment by bradymiller** [comment]
just placed PR proof of concept for encryption of patient documents here:
https://github.com/openemr/openemr/pull/2238

@prondubuisi , let me know if need more guidance on the letters stuff

**Comment by bradymiller** [comment]
document module now supports encryption on drive: https://github.com/openemr/openemr/commit/a960ad8867dc0b6f23aa1a9fe323b06403e1ffb9

**Comment by prondubuisi** [comment]
@bradymiller Frankly am completely lost with how to get started with this task. I am willing to follow through if I get any directives
 though

**Comment by bradymiller** [comment]
hi @prondubuisi ,
Yep, looked through that script and is complicated. Let me see if can provide more specific directives; there's a risk I may end up completing it while figuring it out :) (don't worry, though, there are more places where we will need to support encrypt/decrypt on the drive :) ) -brady

**Comment by bradymiller** [comment]
yep, ended up being complicated but at the same time just requiring a couple lines of code. Here's the code:
https://github.com/openemr/openemr/issues/2181
I'm gonna start to list of other miscellaneous stuff on the top post in this PR to track progress. Guessing there will be up to 10 items or so to do that folks will be able to work on.

**Comment by bradymiller** [comment]
In terms of security of things stored on drive, also several related projects that are important:

1. Folding all the writable directories in documents directory, which would make installation and security of openemr marked improved(also opens the door to support placing the data stored on drive to be migrated outside the web directory):
https://github.com/openemr/openemr/issues/572
https://github.com/openemr/openemr/pull/673
(note these are going to be very extensive changes since it will involve upgrading changes and project infrastructure changes)

2. Ensuring setup checks write permissions on all subdirectories (vital to ensure openemr actually will work):
https://github.com/openemr/openemr/issues/2206

(hopefully will be able to break this project down into bite sized issues; still analyzing)

**Comment by bradymiller** [comment]
PR for "Folding all the writable directories into the documents directory" is here: https://github.com/openemr/openemr/pull/2254

**Comment by bradymiller** [comment]
The critical pieces are completed and can work on the rest for future 5.0.3 release (so changed milestone tag from 5.0.2 to 5.0.3).

**Comment by bradymiller** [comment]
hi @stephenwaite , Here's the related issue (encrypted stuff in documents drive) that we discussed on the call Saturday :)

**Comment by kojiromike** [comment]
Has the status of this ticket been reviewed recently, as we approach 7.0.4?

**Comment by bradymiller** [comment]
hi @kojiromike ,
Would need another strategy evaluation and analysis since it's been so long. There are still things stored documents/ folder that are not encrypted (was the goal of this issue to get to everything which ended up being overly ambitious :) ). Betting new things have crept up on us over this time that are also stored in unencrypted format.
